File: Visual Studio 2008/VBWinFormDesignerCustomPropertyTab/UC_CustomPropertyTab.vb
'************************************* 模块头 ********************************************\
' 模块名称:	UC_CustomPropertyTab.vb
' 项目:		VBWinFormDesignerCustomPropertyTab
' 版权所有 (c) Microsoft Corporation.
' 
' 
' CustomPropertyTab 例子演示了如何在属性窗口上添加自定义的PropertyTab
' 
' 
' This source is subject to the Microsoft Public License.
' See http://www.microsoft.com/opensource/licenses.mspx#Ms-PL.
' All other rights reserved.
' 
' THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
' EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
' WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
'******************************************************************************************/

Imports System.ComponentModel
Imports System.Windows.Forms.PropertyGridInternal


<PropertyTab(GetType(CustomTab), PropertyTabScope.Component)> _
Public Class UC_CustomPropertyTab

    Private _testProp As String

    <Browsable(False), CustomTabDisplay(True)> _
    Public Property TestProp() As String
        Get
            Return Me._testProp
        End Get
        Set(ByVal value As String)
            Me._testProp = value
        End Set
    End Property
End Class


Public Class CustomTab
    Inherits PropertiesTab

    Public Overrides Function CanExtend(ByVal extendee As Object) As Boolean
        Return TypeOf extendee Is UC_CustomPropertyTab
    End Function

    Public Overrides Function GetProperties(ByVal context As ITypeDescriptorContext, _
                    ByVal component As Object, ByVal attrs As Attribute()) _
                    As PropertyDescriptorCollection
        Return TypeDescriptor.GetProperties(component, _
            New Attribute() {New BrowsableAttribute(False), New CustomTabDisplayAttribute(True)})
    End Function

    Public Overrides ReadOnly Property TabName() As String
        Get
            Return "Custom Tab"
        End Get
    End Property
End Class


<AttributeUsage(AttributeTargets.Property)> _
Public Class CustomTabDisplayAttribute
    Inherits Attribute

    Private _display As Boolean

    Public Sub New(ByVal display As Boolean)
        Me.display = display
    End Sub

    Public Property Display() As Boolean
        Get
            Return Me._display
        End Get
        Set(ByVal value As Boolean)
            Me._display = value
        End Set
    End Property
End Class

File: Visual Studio 2008/CSWinFormDesignerCustomPropertyTab/UC_CustomPropertyTab.cs
/************************************* 模块头 *******************************************\
* 模块名称:	UC_CustomPropertyTab.cs
* 项目:		CSWinFormDesignerCustomPropertyTab
* 版权所有 (c) Microsoft Corporation.
* 
* 
* CustomPropertyTab 例子演示了如何在属性窗口上添加自定义的PropertyTab
* 
* 
* This source is subject to the Microsoft Public License.
* See http://www.microsoft.com/opensource/licenses.mspx#Ms-PL.
* All other rights reserved.
* 
* THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
* EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
* WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
\******************************************************************************************/

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Drawing;
using System.Data;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using System.Windows.Forms.PropertyGridInternal;

namespace CSWinFormDesignerCustomPropertyTab
{
    [PropertyTab(typeof(CustomTab), PropertyTabScope.Component)]
    public partial class UC_CustomPropertyTab : UserControl
    {
        private string testProp;

        public UC_CustomPropertyTab()
        {
            InitializeComponent();
        }

        // 显示在自定义选项卡上
        [Browsable(false)]
        [CustomTabDisplayAttribute(true)]
        public string TestProp
        {
            get { return this.testProp; }
            set { this.testProp = value; }
        }
    }

    public class CustomTab : PropertiesTab
    {
        public override bool CanExtend(object extendee)
        {
            // 如果我们扩展这个控件，就返回true            
            return extendee is UC_CustomPropertyTab;
        }

        public override PropertyDescriptorCollection GetProperties(
            ITypeDescriptorContext context,
            object component,
            Attribute[] attrs)
        {
            // 只返回那些被标记为Browserable(false)和CustomTabDisplay(true)的属性
            return TypeDescriptor.GetProperties(
                component,
                new Attribute[] 
                { 
                    new BrowsableAttribute(false), 
                    new CustomTabDisplayAttribute(true) 
                });
        }

        public override string TabName
        {
            get { return "Custom Tab"; }
        }
    }

    [AttributeUsage(AttributeTargets.Property)]
    public class CustomTabDisplayAttribute : Attribute
    {
        private bool display;

        public CustomTabDisplayAttribute(bool display)
        {
            this.display = display;
        }

        public bool Display
        {
            get { return this.display; }
            set { this.display = value; }
        }
    }
}
